### DISCUSSION TITLE
[Chapter 7] UnmodifiableList

### DISCUSSION THREAD

**Discussion by firewoody237** [cowaine/is-my-code-that-strange#63] [Chapter 07-08 (컬렉션, 강한 결합)] [answered]
Chapter 7을 공부하며, `UnmodifiableList`에 대해 처음 알았습니다.(혹은 배웠는데 까먹었거나..)

[링크](https://colabear754.tistory.com/185)제가 참고한 사이트인데, `add()`, `set()`, `remove()` 등을 호출하면 `UnsupportedOperationException`을 던진다고 합니다.
![image](https://github.com/cowaine/is-my-code-that-strange/assets/113076211/a062c0d7-65eb-4a93-8773-5dcba396f690)

다만 원본객체를 복사하는 것이 아닌, 원본객체의 주소를 그대로 가지고 변경에 대한 메서드만 막아주는 형태인데요,
이러한 방법은 일급 컬렉션을 사용해서 캡슐화(내부 리스트가 캡슐화되어 외부에서 절대 참조할 수 없는 형태)를 할 때만 좋은 방식이지 않을까? 하는 의문이 들었습니다.

또한 내부에 참조형태로 되어있다면.. 이 또한 얕은 복사처럼 데이터 자체는 안전하지 않다고 생각했습니다.
아래는 위 생각에 대한 예시코드입니다.

![image](https://github.com/cowaine/is-my-code-that-strange/assets/113076211/90df3358-cbe9-49fd-b628-f9f524163c70)

- 여러분들은 `UnmodifiableXXX`를 사용해보신 적이 있으실까요?
- 결국에 네이밍(Unmodifiable)처럼 완벽하지 않아서, 저는 잘 사용하지 않을 것 같은데.. 여러분의 생각은 어떠신가요?

**Comment by corock** [comment, marked as answer]
> 이러한 방법은 일급 컬렉션을 사용해서 캡슐화(내부 리스트가 캡슐화되어 외부에서 절대 참조할 수 없는 형태)를 할 때만 좋은 방식이지 않을까? 하는 의문이 들었습니다.

(질문에 대한 답변을 작성하기 전에...) 이 생각까지는 못 해봤는데 일급 컬렉션과 같이 사용하면 좋을 것 같네요!

> Q1. 여러분들은 UnmodifiableXXX를 사용해보신 적이 있으실까요?

작년에 창고 관리 시스템(_Warehouse Management Syetem_, 이하 WMS) 전환이 있었는데,
주문 대상 상품에 대해 WMS 로 실시간 재고가 몇 개 있는지 조회하는 API 를 개발한 적이 있습니다.

예를 들어, 서로 다른 주문 대상 상품이 5개가 있다면 각 상품은 WMS 에서 관리하는 SKU(_Stock Keeping Unit_) 코드가 부여되어 있습니다.
클라이언트는 SKU 목록을 요청 값으로 전달해서 재고 조회를 시도하는데 그중 유효하지 않은 SKU 가 있다면 그 개수만큼 응답을 받지 못했습니다.
AS-IS 는 유효하지 않은 응답을 받을 경우 가용 재고를 0으로 처리했는데, 이 과정을 처리할 때 코드 개선하면서 사용해 봤던 기억이 있습니다.

> Q2. 결국에 네이밍(Unmodifiable)처럼 완벽하지 않아서, 저는 잘 사용하지 않을 것 같은데.. 여러분의 생각은 어떠신가요?

위에서 `UnmodifiableList` 클래스로 만든 컬렉션을 최종 반환할 컬렉션에 합치기 위해 `addAll` 메서드를 사용했다가
`UnsupportedOperationException` 예외가 발생해서 그 이후 웬만하면 사용하지 않고 있습니다.
